### DISCUSSION TITLE
Operand Out of Range Error when Compiling ChipKIT-Bootloader

### DISCUSSION THREAD

**Discussion by AlperenAkkuncu** [chipKIT32/chipKIT-core#489] [Q&A] [answered]
Hello,

I've been trying to compile the chipKIT-Bootloaders project on MPLAB X.

I am only interested in one configuration which is UDB32_MX2_DIP. When I try to compile it, MPLAB gives me the following error:

**../crt0.S: Assembler messages:
../crt0.S:261: Error: operand 3 out of range `addiu $10,$0,0xFFFFFFFC'
make[2]: *** [build/UDB32_MX2_DIP/debug/_ext/1472/crt0.o] Error 255
make[2]: *** Waiting for unfinished jobs....**

The main thing I am trying to accomplish is that I am trying to reverse the RX and TX pin on the UART for bootloading the code and for that I need to change re assign the RX and TX pins, and generate hex file.

Is there a way to surpass this error?

PS: I also posted the same question [[here](https://github.com/chipKIT32/chipKIT-digiboot/issues/2)], I don't know which platform is the best for this discussion, but I close it if needed.

**Comment by JacobChrist** [comment]
I recently built a chipKIT bootloader for the PONTECH QUICK240 without issue.  You need to lease or own the XC32 compiler, it will not build without a license.  I believe there are several bootloader projects and the latest is the one listed below.

REPO: https://github.com/chipKIT32/PIC32-avrdude-bootloader.git
GIT SHA-1: e2f14324601fb91f61b7de5810c733efd4eddee2
BUILD DATE: 2023-11-13
COMPILER VERSION: 1.33
BUILD COMPUTER: WINDOWS 11 PRO
TARGET: QUICK240

**Reply by AlperenAkkuncu** [reply]
Thank you for the response Jacob.
Are you certain this is because of license issue? 

I will try to see if it works. If it doesn't work maybe I can request a refund.

**Reply by AlperenAkkuncu** [reply]
I get the same error after installing the PRO version.

I do not know how to solve it really. Have you tried compiling UDB32_MX2_DIP?

Thanks!

**Comment by majenkotech** [comment]
The problem here is that the compiler is more strict in newer versions about operands to 16 bit operations. The instruction, as written, has a 32-bit operand to a 16 bit instruction (`addiu` - ADD Immediate Unsigned). Older versions of the compiler didn't care, or just warned at worst. Newer ones throw an error.

The best fix is to simply truncate the operand (which is what the compiler used to do) by lopping off the first for F's.

```
addiu $10,$0,0xFFFC
```

There will probably be other instances of the same thing though once that one line has been fixed...

Personally I would recommend compiling using the chipKIT compiler not XC32, and doing it under Linux using the included Makefile.

**Reply by AlperenAkkuncu** [reply]
Thank you for your response @majenkotech 

I am not much of a firmware developer, this assembly stuff goes right over my head honestly. 

Do you think I can compile with older versions of XC32 or MPLABX?

**Comment by majenkotech** [comment]
I have pushed a potential fix to the repo. Update and try compiling again (note: I never use XC32 so don't have it set up for testing...)

**Reply by AlperenAkkuncu** [reply]
Thank you very much!

Unfortunately, that didn't work for me as well.

**Here is another idea:**
I don't really need to modify the firmware a lot. I only need to swap RX and TX of the UART that is used to send data. Could you maybe swap the UART lines, compile it and send the hex for me?

Here is the error log in case you want to look at it:

_make -f nbproject/Makefile-UDB32_MX2_DIP.mk SUBPROJECTS= .build-conf
make[1]: Entering directory 'C:/kuncu/Clients/NONAME/Firmware/PIC32-avrdude-bootloader-master/bootloaders/chipKIT-Bootloaders.X'
make  -f nbproject/Makefile-UDB32_MX2_DIP.mk dist/UDB32_MX2_DIP/debug/chipKIT-Bootloaders.X.debug.elf
make[2]: Entering directory 'C:/kuncu/Clients/NONAME/Firmware/PIC32-avrdude-bootloader-master/bootloaders/chipKIT-Bootloaders.X'
"C:\Program Files\Microchip\xc32\v4.35\bin\xc32-gcc.exe"   -g -D__DEBUG   -fframe-base-loclist  -x c -c -mprocessor=32MX250F128B  -mips16 -mno-float -O3 -fno-common -D_BOARD_UDB32_MX2_DIP_ -I".." -MP -MMD -MF "build/UDB32_MX2_DIP/debug/_ext/1472/main.o.d" -o build/UDB32_MX2_DIP/debug/_ext/1472/main.o ../main.c    -DXPRJ_UDB32_MX2_DIP=UDB32_MX2_DIP        
"C:\Program Files\Microchip\xc32\v4.35\bin\xc32-gcc.exe"  -g   -mprocessor=32MX250F128B  -nostdlib -nostartfiles -mips16 -o dist/UDB32_MX2_DIP/debug/chipKIT-Bootloaders.X.debug.elf build/UDB32_MX2_DIP/debug/_ext/1472/main.o build/UDB32_MX2_DIP/debug/_ext/1472/crt0.o          -DXPRJ_UDB32_MX2_DIP=UDB32_MX2_DIP          -Wl,--defsym=__MPLAB_BUILD=1,--script="..\MX1-2-boot-linkerscript.ld",--defsym=__MPLAB_DEBUG=1,--defsym=__DEBUG=1,-D=__DEBUG_D,--no-code-in-dinit,--no-dinit-in-serial-mem,-L"../../../../../../../../Program Files (x86)/Microchip/MPLAB C32 Suite/lib",-L"../../../../../../../../Program Files (x86)/Microchip/MPLAB C32 Suite/pic32mx/lib",-Map="dist/UDB32_MX2_DIP/debug/chipKIT-Bootloaders.X.debug.map",--memorysummary,dist/UDB32_MX2_DIP/debug/memoryfile.xml,-Os 
nbproject/Makefile-UDB32_MX2_DIP.mk:165: recipe for target 'dist/UDB32_MX2_DIP/debug/chipKIT-Bootloaders.X.debug.elf' failed
make[2]: Leaving directory 'C:/kuncu/Clients/NONAME/Firmware/PIC32-avrdude-bootloader-master/bootloaders/chipKIT-Bootloaders.X'
nbproject/Makefile-UDB32_MX2_DIP.mk:101: recipe for target '.build-conf' failed
make[1]: Leaving directory 'C:/kuncu/Clients/NONAME/Firmware/PIC32-avrdude-bootloader-master/bootloaders/chipKIT-Bootloaders.X'
nbproject/Makefile-impl.mk:39: recipe for target '.build-impl' failed
Warning: dinit optimization threshholds not set; dinit may not function
build/UDB32_MX2_DIP/debug/_ext/1472/main.o: Link Error: Could not allocate section .text, size = 3464 bytes, attributes = code 
 Link Error: Could not allocate program memory

Info: Loading file: ./..\MX1-2-boot-linkerscript.ld
collect2.exe: error: ld returned 255 exit status
make[2]: *** [dist/UDB32_MX2_DIP/debug/chipKIT-Bootloaders.X.debug.elf] Error 255
make[1]: *** [.build-conf] Error 2
make: *** [.build-impl] Error 2_

**Comment by majenkotech** [comment]
Sure, if you can drop the UARTMapRX and UARTMapTX entries on here to save me hunting through the datasheet for the right values I can do that for you.

**Reply by AlperenAkkuncu** [reply]
I just checked the datasheet. I think it's not possible to do that... very dissapointed. I thought these parts were very flexible, but apparently not.

I need RPB3 to be U1RX and RPB13 to be U1TX, but it seems like it's not possible if I am not missing anything. 

If you have time, you can check as well.

**Comment by majenkotech** [comment, marked as answer]
Yeah they're not a complete all:all matrix - only select pins can do select functions. There's no UART RX functionality on B3, and no UART TX functionality on B13. Time for a re-think?

**Reply by AlperenAkkuncu** [reply]
Time for fixing the board with wires and solder.

It seems like fix is not possible in the firmware level. 

Thank you for your help :)
